checkup: Add guest agent ping wait

RamLavi opened the pull request:
This PR introduces a new algorithm to make sure that the checkups start only after the vmi properly configured for DPDK.
Instead of waiting for the VMI guest-agent condition to be true, introducing a guest-agent-ping probe that will poll for the marker file signaling that the VMI is ready.

~~It relies on #220, first 4 commits do not belong to this PR~~

orelmisan reviewed (commented):
Thank you for the PR @RamLavi.
1. Isn't there a need to use `chcon -t virt_qemu_ga_exec_t <marker_file_name>` for the probe to work?
2. Please consider replacing the GA connected condition with the ready condition, as the ready condition is stronger and contains GA connected.

orelmisan commented on pkg/internal/checkup/vmi/spec.go:
```diff
@@ -129,6 +129,22 @@ func WithVirtIODisk(name string) Option {
 	}
 }
 
+func WithReadinessFileProbe(fileName string) Option {
```
Could you please move it to the end of the file?

RamLavi commented:
e2e passes on a CNV 4.15 cluster:
```
make test/e2e
podman run --rm \
           --volume /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup:/home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup:Z \
           --volume /home/ralavi/.kube/sno01-cnvqe2-rdu2:/root/.kube:Z,ro \
           --workdir /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup \
           -e KUBECONFIG=/root/.kube/kubeconfig \
           -e TEST_CHECKUP_IMAGE=quay.io/ramlavi/kubevirt-dpdk-checkup:latest \
           -e TEST_NAMESPACE=dpdk-checkup-ns-1 \
           -e NETWORK_ATTACHMENT_DEFINITION_NAME=dpdk-sriovnetwork-ns-1 \
           -e TRAFFIC_GEN_CONTAINER_DISK_IMAGE=quay.io/ramlavi/kubevirt-dpdk-checkup-traffic-gen:latest \
           -e VM_UNDER_TEST_CONTAINER_DISK_IMAGE=quay.io/ramlavi/kubevirt-dpdk-checkup-vm:latest \
           docker.io/library/golang:1.20.12 go test ./tests/... -test.v -test.timeout=1h -ginkgo.v -ginkgo.timeout=1h
=== RUN   TestKubevirtDpdkCheckup
Running Suite: KubevirtDpdkCheckup Suite - /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests
==============================================================================================================
Random Seed: 1706193086

Will run 1 of 1 specs
------------------------------
[BeforeSuite] 
/home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/test_suite_test.go:56
[BeforeSuite] PASSED [0.002 seconds]
------------------------------
Execute the checkup Job should complete successfully
/home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:83
• [335.643 seconds]
------------------------------

Ran 1 of 1 Specs in 335.645 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestKubevirtDpdkCheckup (335.65s)
PASS
ok  	github.com/kiagnose/kubevirt-dpdk-checkup/tests
```

logs:
```
2024/01/25 14:31:34 kubevirt-dpdk-checkup starting...
2024/01/25 14:31:34 Using the following config:
2024/01/25 14:31:34 "timeout": "1h0m0s"
2024/01/25 14:31:34 "networkAttachmentDefinitionName": "dpdk-sriovnetwork-ns-1"
2024/01/25 14:31:34 "trafficGenContainerDiskImage": "quay.io/ramlavi/kubevirt-dpdk-checkup-traffic-gen:latest"
2024/01/25 14:31:34 "trafficGenTargetNodeName": ""
2024/01/25 14:31:34 "trafficGenPacketsPerSecond": "8m"
2024/01/25 14:31:34 "TrafficGenEastMacAddress": "50:e1:a0:87:5d:01"
2024/01/25 14:31:34 "TrafficGenWestMacAddress": "50:5c:bf:14:ec:02"
2024/01/25 14:31:34 "vmUnderTestContainerDiskImage": "quay.io/ramlavi/kubevirt-dpdk-checkup-vm:latest"
2024/01/25 14:31:34 "vmUnderTestTargetNodeName": ""
2024/01/25 14:31:34 "VMUnderTestEastMacAddress": "60:2b:1f:0c:d5:01"
2024/01/25 14:31:34 "VMUnderTestWestMacAddress": "60:44:3c:14:52:02"
2024/01/25 14:31:34 "testDuration": "1s"
2024/01/25 14:31:34 "portBandwidthGbps": "10"
2024/01/25 14:31:34 "verbose": true
2024/01/25 14:31:34 Creating ConfigMap "dpdk-checkup-ns-1/dpdk-traffic-gen-config-5842m"...
2024/01/25 14:31:34 Creating ConfigMap "dpdk-checkup-ns-1/vmi-under-test-config-5842m"...
2024/01/25 14:31:34 Creating VMI "dpdk-checkup-ns-1/vmi-under-test-5842m"...
2024/01/25 14:31:34 Creating VMI "dpdk-checkup-ns-1/dpdk-traffic-gen-5842m"...
2024/01/25 14:31:34 Waiting for VMI "dpdk-checkup-ns-1/vmi-under-test-5842m" ready condition...
2024/01/25 14:34:14 VMI "dpdk-checkup-ns-1/vmi-under-test-5842m" has successfully reached ready condition
2024/01/25 14:34:14 Waiting for VMI "dpdk-checkup-ns-1/dpdk-traffic-gen-5842m" ready condition...
2024/01/25 14:35:19 VMI "dpdk-checkup-ns-1/dpdk-traffic-gen-5842m" has successfully reached ready condition
2024/01/25 14:35:19 Login to VMI under test...
2024/01/25 14:35:47 Login to traffic generator...
2024/01/25 14:35:55 VMI under test guest kernel Args: cat /proc/cmdline
BOOT_IMAGE=(hd0,gpt2)/vmlinuz-4.18.0-365.el8.x86_64 root=UUID=a9332d7d-1762-41cd-a702-6b2cc556c248 ro console=tty0 rd_NO_PLYMOUTH crashkernel=auto resume=UUID=d4ac9572-e828-4c87-9b76-e59c9fa6e426 console=ttyS0,115200 default_hugepagesz=1GB hugepagesz=1G hugepages=1 skew_tick=1 tsc=reliable rcupdate.rcu_normal_after_boot=1 nohz=on nohz_full=2-7 rcu_nocbs=2-7 tuned.non_isolcpus=00000003 intel_pstate=disable nosoftlockup
[root@vmi-under-test-5842m cloud-user]# 
2024/01/25 14:35:57 traffic generator guest kernel Args: cat /proc/cmdline
BOOT_IMAGE=(hd0,gpt2)/vmlinuz-4.18.0-365.el8.x86_64 root=UUID=a9332d7d-1762-41cd-a702-6b2cc556c248 ro console=tty0 rd_NO_PLYMOUTH crashkernel=auto resume=UUID=d4ac9572-e828-4c87-9b76-e59c9fa6e426 console=ttyS0,115200 default_hugepagesz=1GB hugepagesz=1G hugepages=1
[root@dpdk-traffic-gen-5842m cloud-user]# 
2024/01/25 14:35:57 Starting traffic generator Server Service...
2024/01/25 14:35:57 Waiting until traffic generator Server Service is ready...
2024/01/25 14:36:09 trex-console help resp:
cd /opt/trex && echo "help" | ./trex-console

Using 'python3' as Python interpeter

Log:
*** [RPC] - Failed to get server response from tcp://localhost:4501

[root@dpdk-traffic-gen-5842m trex]# 
2024/01/25 14:36:09 trex-server is not yet ready...
2024/01/25 14:36:18 trex-console help resp:
cd /opt/trex && echo "help" | ./trex-console

Using 'python3' as Python interpeter


Connecting to RPC server on localhost:4501                   [SUCCESS]


Connecting to publisher server on localhost:4500             [SUCCESS]


Acquiring ports [0, 1]:                                      [SUCCESS]

*** Warning - Port 0 destination is unresolved ***
*** Warning - Port 1 destination is unresolved ***

Server Info:

Server version:   v3.03 @ STL
Server mode:      Stateless
Server CPU:       4 x Intel Xeon Processor (Cascadelake)
Ports count:      2 x 10.0Gbps @ Ethernet Virtual Function 700 Series	

-=TRex Console v3.0=-

Type 'help' or '?' for supported actions

trex>
Console Commands:

capture -                      Manage PCAP captures          
debug -                        Internal debugger for development.
events -                       Shows events log              
help -                         Shows This Help Screen        
history -                      Manage the command history    
plugins -                      Show / load / use plugins     
quit -                         Exit the console              
tui -                          Shows a graphical console     
verbose -                      Shows or set verbose mode     

Common Commands:

acquire -                      Acquire ports                 
arp -                          Performs a port ARP resolution
clear -                        Clear cached local statistics 
connect -                      Connects to the TRex server and acquire ports
disconnect -                   Disconnect from the TRex server
global_cfg -                   Set global configuration parameters
ipv6 -                         Configures IPv6 of a port     
l2 -                           Configures a port in L2 mode  
l3 -                           Configures a port in L3 mode  
map -                          Maps ports topology           
ns -                           Network namespace             
ping -                         Pings the server / specific IP
pkt -                          Sends a Scapy format packet   
portattr -                     Sets port attributes          
release -                      Release ports                 
reset -                        Reset ports                   
scan6 -                        Search for IPv6 neighbors     
shutdown -                     Shutdown the server           
stats -                        Show various statistics       
vlan -                         Configures VLAN tagging for a port.

Stateless Commands:

pause -                        Pause active traffic on specified ports on TRex
profiles -                     Get loaded to server profiles information
push -                         Push a pcap file              
resume -                       Resume active traffic on specified ports on TRex
service -                      Configures port for service mode.
start -                        Start selected traffic on specified ports on TRex
stop -                         Stop active traffic on specified ports on TRex
streams -                      Get loaded to server streams information
tpg_clear_stats -              Clear Tagged Packet Group Stats
tpg_clear_tx_stats -           Clear Tagged Packet Group Tx Stats
tpg_clear_unknown_tags -       Clear Tagged Packet Group Unknown Tags
tpg_disable -                  Disable Tagged Packet Group   
tpg_enable -                   Enable Tagged Packet Group    
tpg_show_unknown_tags -        Show Tagged Packet Group Unknown Tags
tpg_stats -                    Show Tagged Packet Group Statistics
tpg_status -                   Show Tagged Packet Group Status
tpg_tags -                     Show Tagged Packet Group Tags 
tpg_tx_stats -                 Show Tagged Packet Group Tx Statistics
tpg_update -                   Update Tagged Packet Group Tag
update -                       Update port(s) speed currently active
trex>Shutting down RPC client

[root@dpdk-traffic-gen-5842m trex]# 
2024/01/25 14:36:18 trex-server is now ready
2024/01/25 14:36:18 Starting testpmd in VMI...
2024/01/25 14:36:24 testpmd run:
dpdk-testpmd --lcores 0@2-3,1@4,2@5,3@6,4@7 -a 0000:06:00.0 -a 0000:07:00.0 --socket-mem 1024 --huge-dir /mnt/huge -- -i --nb-cores=4 --rxd=2048 --txd=2048 --rxq=4 --txq=4 --forward-mode=mac --eth-peer=0,50:e1:a0:87:5d:01 --eth-peer=1,50:5c:bf:14:ec:02
EAL: Detected CPU lcores: 8
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: No available 2048 kB hugepages reported
EAL: VFIO support initialized
EAL: Using IOMMU type 8 (No-IOMMU)
EAL: Probe PCI driver: net_iavf (8086:154c) device: 0000:06:00.0 (socket 0)
EAL: Probe PCI driver: net_iavf (8086:154c) device: 0000:07:00.0 (socket 0)
TELEMETRY: No legacy callbacks, legacy socket not created
Interactive-mode selected
Set mac packet forwarding mode
testpmd: create a new mbuf pool <mb_pool_0>: n=179456, size=2176, socket=0
testpmd: preferred mempool ops selected: ring_mp_mc
Configuring Port 0 (socket 0)
iavf_dev_init_vlan(): Failed to update vlan offload
iavf_dev_configure(): configure VLAN failed: -95
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[0]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[1]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[2]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[3]

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event
Port 0: 60:2B:1F:0C:D5:01
Configuring Port 1 (socket 0)
iavf_dev_init_vlan(): Failed to update vlan offload
iavf_dev_configure(): configure VLAN failed: -95
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[0]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[1]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[2]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[3]

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event
Port 1: 60:44:3C:14:52:02
Checking link statuses...
Done
testpmd> 
2024/01/25 14:36:24 testpmd start:
start
mac packet forwarding - ports=2 - cores=4 - streams=8 - NUMA support enabled, MP allocation mode: native
Logical Core 1 (socket 0) forwards packets on 2 streams:
  RX P=0/Q=0 (socket 0) -> TX P=1/Q=0 (socket 0) peer=50:5C:BF:14:EC:02
  RX P=1/Q=0 (socket 0) -> TX P=0/Q=0 (socket 0) peer=50:E1:A0:87:5D:01
Logical Core 2 (socket 0) forwards packets on 2 streams:
  RX P=0/Q=1 (socket 0) -> TX P=1/Q=1 (socket 0) peer=50:5C:BF:14:EC:02
  RX P=1/Q=1 (socket 0) -> TX P=0/Q=1 (socket 0) peer=50:E1:A0:87:5D:01
Logical Core 3 (socket 0) forwards packets on 2 streams:
  RX P=0/Q=2 (socket 0) -> TX P=1/Q=2 (socket 0) peer=50:5C:BF:14:EC:02
  RX P=1/Q=2 (socket 0) -> TX P=0/Q=2 (socket 0) peer=50:E1:A0:87:5D:01
Logical Core 4 (socket 0) forwards packets on 2 streams:
  RX P=0/Q=3 (socket 0) -> TX P=1/Q=3 (socket 0) peer=50:5C:BF:14:EC:02
  RX P=1/Q=3 (socket 0) -> TX P=0/Q=3 (socket 0) peer=50:E1:A0:87:5D:01

  mac packet forwarding packets/burst=32
  nb forwarding cores=4 - nb forwarding ports=2
  port 0: RX queue number: 4 Tx queue number: 4
    Rx offloads=0x80000 Tx offloads=0x10000
    RX queue: 0
      RX desc=2048 - RX free threshold=32
      RX threshold registers: pthresh=0 hthresh=0  wthresh=0
      RX Offloads=0x0
    TX queue: 0
      TX desc=2048 - TX free threshold=32
      TX threshold registers: pthresh=0 hthresh=0  wthresh=0
      TX offloads=0x10000 - TX RS bit threshold=32
  port 1: RX queue number: 4 Tx queue number: 4
    Rx offloads=0x80000 Tx offloads=0x10000
    RX queue: 0
      RX desc=2048 - RX free threshold=32
      RX threshold registers: pthresh=0 hthresh=0  wthresh=0
      RX Offloads=0x0
    TX queue: 0
      TX desc=2048 - TX free threshold=32
      TX threshold registers: pthresh=0 hthresh=0  wthresh=0
      TX offloads=0x10000 - TX RS bit threshold=32
testpmd> 
2024/01/25 14:36:24 Clearing testpmd stats in VMI...
2024/01/25 14:36:24 Clearing Trex console stats before test...
2024/01/25 14:36:26 Running traffic for 1s...
2024/01/25 14:36:28 Monitoring traffic generator side drop rates every 10s during the test duration...
2024/01/25 14:36:30 GetGlobalStats JSON Response:
{
    "id": "rnxxazwr",
    "jsonrpc": "2.0",
    "result": {
        "m_active_flows": 0.0,
        "m_active_sockets": 0,
        "m_bw_per_core": 82.27063751220703,
        "m_cpu_util": 1.213937520980835,
        "m_cpu_util_raw": 4.25,
        "m_open_flows": 0.0,
        "m_platform_factor": 1.0,
        "m_rx_bps": 1868076672.0,
        "m_rx_core_pps": 0.0,
        "m_rx_cpu_util": 0.0,
        "m_rx_drop_bps": 0.0,
        "m_rx_pps": 3648587.25,
        "m_socket_util": 0.0,
        "m_total_alloc_error": 0,
        "m_total_clients": 0,
        "m_total_nat_active": 0,
        "m_total_nat_learn_error": 0,
        "m_total_nat_no_fid": 0,
        "m_total_nat_open": 0,
        "m_total_nat_syn_wait": 0,
        "m_total_nat_time_out": 0,
        "m_total_nat_time_out_wait_ack": 0,
        "m_total_queue_drop": 0,
        "m_total_queue_full": 0,
        "m_total_rx_bytes": 512000256,
        "m_total_rx_pkts": 8000004,
        "m_total_servers": 0,
        "m_total_tx_bytes": 544000272,
        "m_total_tx_pkts": 8000004,
        "m_tx_bps": 1997428224.0,
        "m_tx_cps": 0.0,
        "m_tx_expected_bps": 0.0,
        "m_tx_expected_cps": 0.0,
        "m_tx_expected_pps": 0.0,
        "m_tx_pps": 3671743.0
    }
}
2024/01/25 14:36:30 finished polling for drop rates
2024/01/25 14:36:30 traffic Generator Max Drop Rate: 0.000000Bps
2024/01/25 14:36:32 GetPortStats JSON Response:
{
    "id": "7qk64crr",
    "jsonrpc": "2.0",
    "result": {
        "ibytes": 0,
        "ierrors": 0,
        "ipackets": 0,
        "m_cpu_util": 0.0,
        "m_total_rx_bps": 0.0,
        "m_total_rx_pps": 0.0,
        "m_total_tx_bps": 998714112.0,
        "m_total_tx_pps": 1835871.5,
        "obytes": 544000272,
        "oerrors": 0,
        "opackets": 8000004
    }
}
2024/01/25 14:36:33 GetPortStats JSON Response:
{
    "id": "329hgdsl",
    "jsonrpc": "2.0",
    "result": {
        "ibytes": 512000256,
        "ierrors": 0,
        "ipackets": 8000004,
        "m_cpu_util": 0.0,
        "m_total_rx_bps": 233509584.0,
        "m_total_rx_pps": 456073.40625,
        "m_total_tx_bps": 0.0,
        "m_total_tx_pps": 0.0,
        "obytes": 0,
        "oerrors": 0,
        "opackets": 0
    }
}
2024/01/25 14:36:33 traffic Generator port 0 Packet output errors: 0
2024/01/25 14:36:33 traffic Generator port 1 Packet output errors: 0
2024/01/25 14:36:33 traffic Generator packet sent via port 0: 8000004
2024/01/25 14:36:33 get testpmd stats in VM-Under-Test...
2024/01/25 14:36:36 testpmd stats:
show fwd stats all

  ------- Forward Stats for RX Port= 0/Queue= 0 -> TX Port= 1/Queue= 0 -------
  RX-packets: 4000002        TX-packets: 4000002        TX-dropped: 0             

  ------- Forward Stats for RX Port= 0/Queue= 1 -> TX Port= 1/Queue= 1 -------
  RX-packets: 4000002        TX-packets: 4000002        TX-dropped: 0             

  ---------------------- Forward statistics for port 0  ----------------------
  RX-packets: 8000004        RX-dropped: 0             RX-total: 8000004
  TX-packets: 0              TX-dropped: 0             TX-total: 0
  ----------------------------------------------------------------------------

  ---------------------- Forward statistics for port 1  ----------------------
  RX-packets: 0              RX-dropped: 0             RX-total: 0
  TX-packets: 8000004        TX-dropped: 0             TX-total: 8000004
  ----------------------------------------------------------------------------

  +++++++++++++++ Accumulated forward statistics for all ports+++++++++++++++
  RX-packets: 8000004        RX-dropped: 0             RX-total: 8000004
  TX-packets: 8000004        TX-dropped: 0             TX-total: 8000004
  ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
testpmd> 
2024/01/25 14:36:36 VMI-Under-Test's side packets Dropped: Rx: 0; TX: 0
2024/01/25 14:36:36 VMI-Under-Test's side test packets received (including dropped, excluding non-related packets): 8000004
2024/01/25 14:36:36 Trying to delete VMI: "dpdk-checkup-ns-1/vmi-under-test-5842m"
2024/01/25 14:36:39 Trying to delete VMI: "dpdk-checkup-ns-1/dpdk-traffic-gen-5842m"
2024/01/25 14:36:39 Deleting ConfigMap "dpdk-checkup-ns-1/dpdk-traffic-gen-config-5842m"...
2024/01/25 14:36:39 Deleting ConfigMap "dpdk-checkup-ns-1/vmi-under-test-config-5842m"...
2024/01/25 14:36:39 Waiting for VMI "dpdk-checkup-ns-1/vmi-under-test-5842m" to be deleted...
2024/01/25 14:36:54 VMI "dpdk-checkup-ns-1/vmi-under-test-5842m" was deleted successfully
2024/01/25 14:36:54 Waiting for VMI "dpdk-checkup-ns-1/dpdk-traffic-gen-5842m" to be deleted...
2024/01/25 14:36:54 VMI "dpdk-checkup-ns-1/dpdk-traffic-gen-5842m" was deleted successfully
```

RamLavi commented on pkg/internal/checkup/checkup.go:
```diff
@@ -233,6 +232,15 @@ func (c *Checkup) createVMI(ctx context.Context, vmiToCreate *kvcorev1.VirtualMa
 	return err
 }
 
+func (c *Checkup) setupVMIWaitReady(ctx context.Context, name string) (*kvcorev1.VirtualMachineInstance, error) {
```
we also want to do the soft reset there..

orelmisan replied:
The soft reboot is only introduced in #222.
I suggest to introduce it there if you find it necessary.

orelmisan commented:
Quick idea: What would happen if we will have two files
One to mark that tuned-adm was applied, and another to mark the VM readyness after the reboot.
This way we could still have the reboot in the script (instead of an API call), and catch the readiness state only once.
WDYT?

orelmisan commented on pkg/internal/checkup/vmi/spec.go:
```diff
@@ -262,6 +262,22 @@ func WithAffinity(affinity *corev1.Affinity) Option {
+					Command: readinessProbeCommand,
+				},
+			},
+			FailureThreshold:    30,
```
Adding `TimeoutSeconds:      10,` helped reduce the number of queries GA makes, and gets to the Ready state faster.

RamLavi replied:
DONE

RamLavi commented:
[Change](https://github.com/kiagnose/kubevirt-dpdk-checkup/compare/35bbe3000b43a0565b9ce13da0f28995f33f74cd..2748f5667c75631f845613ca2e2fb31af6a6312e): 
- Replaced the "set -e" flag  with "set -x" in the boot script
- Added TimeoutSeconds to the guest-agent-ping probe

orelmisan commented on pkg/internal/checkup/vmi.go:
```diff
@@ -136,7 +136,7 @@ func generateBootScript() string {
 
 	sb.WriteString("#!/bin/bash\n")
-	sb.WriteString("set -e\n")
+	sb.WriteString("set -x\n")
```
Where can one see the output of this script?

orelmisan replied:
@RamLavi Added this info to the commit message.

RamLavi commented:
passed CI on CNV 4.15 cluster:
```
make test/e2e
podman run --rm \
           --volume /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup:/home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup:Z \
           --volume /home/ralavi/.kube/sno01-cnvqe2-rdu2:/root/.kube:Z,ro \
           --workdir /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup \
           -e KUBECONFIG=/root/.kube/kubeconfig \
           -e TEST_CHECKUP_IMAGE=quay.io/ramlavi/kubevirt-dpdk-checkup:latest \
           -e TEST_NAMESPACE=dpdk-checkup-ns-1 \
           -e NETWORK_ATTACHMENT_DEFINITION_NAME=dpdk-sriovnetwork-ns-1 \
           -e TRAFFIC_GEN_CONTAINER_DISK_IMAGE=quay.io/ramlavi/kubevirt-dpdk-checkup-traffic-gen:latest \
           -e VM_UNDER_TEST_CONTAINER_DISK_IMAGE=quay.io/ramlavi/kubevirt-dpdk-checkup-vm:latest \
           docker.io/library/golang:1.20.12 go test ./tests/... -test.v -test.timeout=1h -ginkgo.v -ginkgo.timeout=1h
=== RUN   TestKubevirtDpdkCheckup
Running Suite: KubevirtDpdkCheckup Suite - /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests
==============================================================================================================
Random Seed: 1706437023

Will run 1 of 1 specs
------------------------------
[BeforeSuite] 
/home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/test_suite_test.go:56
[BeforeSuite] PASSED [0.002 seconds]
------------------------------
Execute the checkup Job should complete successfully
/home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:83
• [388.660 seconds]
------------------------------

Ran 1 of 1 Specs in 388.662 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestKubevirtDpdkCheckup (388.66s)
PASS
ok  	github.com/kiagnose/kubevirt-dpdk-checkup/tests	388.681s
```

logs:
```
2024/01/28 10:17:09 kubevirt-dpdk-checkup starting...
2024/01/28 10:17:09 Using the following config:
2024/01/28 10:17:09 "timeout": "1h0m0s"
2024/01/28 10:17:09 "networkAttachmentDefinitionName": "dpdk-sriovnetwork-ns-1"
2024/01/28 10:17:09 "trafficGenContainerDiskImage": "quay.io/ramlavi/kubevirt-dpdk-checkup-traffic-gen:latest"
2024/01/28 10:17:09 "trafficGenTargetNodeName": ""
2024/01/28 10:17:09 "trafficGenPacketsPerSecond": "8m"
2024/01/28 10:17:09 "TrafficGenEastMacAddress": "50:be:76:14:22:01"
2024/01/28 10:17:09 "TrafficGenWestMacAddress": "50:e3:44:47:9b:02"
2024/01/28 10:17:09 "vmUnderTestContainerDiskImage": "quay.io/ramlavi/kubevirt-dpdk-checkup-vm:latest"
2024/01/28 10:17:09 "vmUnderTestTargetNodeName": ""
2024/01/28 10:17:09 "VMUnderTestEastMacAddress": "60:54:27:52:a7:01"
2024/01/28 10:17:09 "VMUnderTestWestMacAddress": "60:54:96:70:df:02"
2024/01/28 10:17:09 "testDuration": "1s"
2024/01/28 10:17:09 "portBandwidthGbps": "10"
2024/01/28 10:17:09 "verbose": true
2024/01/28 10:17:09 Creating ConfigMap "dpdk-checkup-ns-1/dpdk-traffic-gen-config-pnjqb"...
2024/01/28 10:17:09 Creating ConfigMap "dpdk-checkup-ns-1/vmi-under-test-config-pnjqb"...
2024/01/28 10:17:09 Creating VMI "dpdk-checkup-ns-1/vmi-under-test-pnjqb"...
2024/01/28 10:17:09 Creating VMI "dpdk-checkup-ns-1/dpdk-traffic-gen-pnjqb"...
2024/01/28 10:17:09 Waiting for VMI "dpdk-checkup-ns-1/vmi-under-test-pnjqb" to be ready...
2024/01/28 10:21:34 VMI "dpdk-checkup-ns-1/vmi-under-test-pnjqb" has successfully reached ready condition
2024/01/28 10:21:34 Waiting for VMI "dpdk-checkup-ns-1/dpdk-traffic-gen-pnjqb" to be ready...
2024/01/28 10:22:04 VMI "dpdk-checkup-ns-1/dpdk-traffic-gen-pnjqb" has successfully reached ready condition
2024/01/28 10:22:04 Login to VMI under test...
2024/01/28 10:22:13 Login to traffic generator...
2024/01/28 10:22:22 VMI under test guest kernel Args: cat /proc/cmdline
BOOT_IMAGE=(hd0,gpt2)/vmlinuz-4.18.0-365.el8.x86_64 root=UUID=a9332d7d-1762-41cd-a702-6b2cc556c248 ro console=tty0 rd_NO_PLYMOUTH crashkernel=auto resume=UUID=d4ac9572-e828-4c87-9b76-e59c9fa6e426 console=ttyS0,115200 default_hugepagesz=1GB hugepagesz=1G hugepages=1 skew_tick=1 tsc=reliable rcupdate.rcu_normal_after_boot=1 nohz=on nohz_full=2-7 rcu_nocbs=2-7 tuned.non_isolcpus=00000003 intel_pstate=disable nosoftlockup
[root@vmi-under-test-pnjqb ~]# 
2024/01/28 10:22:24 traffic generator guest kernel Args: cat /proc/cmdline
BOOT_IMAGE=(hd0,gpt2)/vmlinuz-4.18.0-365.el8.x86_64 root=UUID=a9332d7d-1762-41cd-a702-6b2cc556c248 ro console=tty0 rd_NO_PLYMOUTH crashkernel=auto resume=UUID=d4ac9572-e828-4c87-9b76-e59c9fa6e426 console=ttyS0,115200 default_hugepagesz=1GB hugepagesz=1G hugepages=1 skew_tick=1 tsc=reliable rcupdate.rcu_normal_after_boot=1 nohz=on nohz_full=2-7 rcu_nocbs=2-7 tuned.non_isolcpus=00000003 intel_pstate=disable nosoftlockup
[root@dpdk-traffic-gen-pnjqb ~]# 
2024/01/28 10:22:24 Starting traffic generator Server Service...
2024/01/28 10:22:24 Waiting until traffic generator Server Service is ready...
2024/01/28 10:22:35 trex-console help resp:
cd /opt/trex && echo "help" | ./trex-console

Using 'python3' as Python interpeter

Log:
*** [RPC] - Failed to get server response from tcp://localhost:4501

[root@dpdk-traffic-gen-pnjqb trex]# 
2024/01/28 10:22:35 trex-server is not yet ready...
2024/01/28 10:22:47 trex-console help resp:
cd /opt/trex && echo "help" | ./trex-console

Using 'python3' as Python interpeter


Connecting to RPC server on localhost:4501                   [SUCCESS]


Connecting to publisher server on localhost:4500             [SUCCESS]


Acquiring ports [0, 1]:                                      [SUCCESS]

*** Warning - Port 0 destination is unresolved ***
*** Warning - Port 1 destination is unresolved ***

Server Info:

Server version:   v3.03 @ STL
Server mode:      Stateless
Server CPU:       4 x Intel Xeon Processor (Cascadelake)
Ports count:      2 x 10.0Gbps @ Ethernet Virtual Function 700 Series	

-=TRex Console v3.0=-

Type 'help' or '?' for supported actions

trex>
Console Commands:

capture -                      Manage PCAP captures          
debug -                        Internal debugger for development.
events -                       Shows events log              
help -                         Shows This Help Screen        
history -                      Manage the command history    
plugins -                      Show / load / use plugins     
quit -                         Exit the console              
tui -                          Shows a graphical console     
verbose -                      Shows or set verbose mode     

Common Commands:

acquire -                      Acquire ports                 
arp -                          Performs a port ARP resolution
clear -                        Clear cached local statistics 
connect -                      Connects to the TRex server and acquire ports
disconnect -                   Disconnect from the TRex server
global_cfg -                   Set global configuration parameters
ipv6 -                         Configures IPv6 of a port     
l2 -                           Configures a port in L2 mode  
l3 -                           Configures a port in L3 mode  
map -                          Maps ports topology           
ns -                           Network namespace             
ping -                         Pings the server / specific IP
pkt -                          Sends a Scapy format packet   
portattr -                     Sets port attributes          
release -                      Release ports                 
reset -                        Reset ports                   
scan6 -                        Search for IPv6 neighbors     
shutdown -                     Shutdown the server           
stats -                        Show various statistics       
vlan -                         Configures VLAN tagging for a port.

Stateless Commands:

pause -                        Pause active traffic on specified ports on TRex
profiles -                     Get loaded to server profiles information
push -                         Push a pcap file              
resume -                       Resume active traffic on specified ports on TRex
service -                      Configures port for service mode.
start -                        Start selected traffic on specified ports on TRex
stop -                         Stop active traffic on specified ports on TRex
streams -                      Get loaded to server streams information
tpg_clear_stats -              Clear Tagged Packet Group Stats
tpg_clear_tx_stats -           Clear Tagged Packet Group Tx Stats
tpg_clear_unknown_tags -       Clear Tagged Packet Group Unknown Tags
tpg_disable -                  Disable Tagged Packet Group   
tpg_enable -                   Enable Tagged Packet Group    
tpg_show_unknown_tags -        Show Tagged Packet Group Unknown Tags
tpg_stats -                    Show Tagged Packet Group Statistics
tpg_status -                   Show Tagged Packet Group Status
tpg_tags -                     Show Tagged Packet Group Tags 
tpg_tx_stats -                 Show Tagged Packet Group Tx Statistics
tpg_update -                   Update Tagged Packet Group Tag
update -                       Update port(s) speed currently active
trex>Shutting down RPC client

[root@dpdk-traffic-gen-pnjqb trex]# 
2024/01/28 10:22:47 trex-server is now ready
2024/01/28 10:22:47 Starting testpmd in VMI...
2024/01/28 10:22:55 testpmd run:
dpdk-testpmd --lcores 0@2-3,1@4,2@5,3@6,4@7 -a 0000:06:00.0 -a 0000:07:00.0 --socket-mem 1024 --huge-dir /mnt/huge -- -i --nb-cores=4 --rxd=2048 --txd=2048 --rxq=4 --txq=4 --forward-mode=mac --eth-peer=0,50:be:76:14:22:01 --eth-peer=1,50:e3:44:47:9b:02
EAL: Detected CPU lcores: 8
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: No available 2048 kB hugepages reported
EAL: VFIO support initialized
EAL: Using IOMMU type 8 (No-IOMMU)
EAL: Probe PCI driver: net_iavf (8086:154c) device: 0000:06:00.0 (socket 0)
EAL: Probe PCI driver: net_iavf (8086:154c) device: 0000:07:00.0 (socket 0)
TELEMETRY: No legacy callbacks, legacy socket not created
Interactive-mode selected
Set mac packet forwarding mode
testpmd: create a new mbuf pool <mb_pool_0>: n=179456, size=2176, socket=0
testpmd: preferred mempool ops selected: ring_mp_mc
Configuring Port 0 (socket 0)
iavf_dev_init_vlan(): Failed to update vlan offload
iavf_dev_configure(): configure VLAN failed: -95
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[0]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[1]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[2]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[3]

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event

Port 0: link state change event
Port 0: 60:54:27:52:A7:01
Configuring Port 1 (socket 0)
iavf_dev_init_vlan(): Failed to update vlan offload
iavf_dev_configure(): configure VLAN failed: -95
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[0]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[1]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[2]
iavf_configure_queues(): RXDID[1] is not supported, request default RXDID[1] in Queue[3]

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event

Port 1: link state change event
Port 1: 60:54:96:70:DF:02
Checking link statuses...
Done
testpmd> 
2024/01/28 10:22:55 testpmd start:
start
mac packet forwarding - ports=2 - cores=4 - streams=8 - NUMA support enabled, MP allocation mode: native
Logical Core 1 (socket 0) forwards packets on 2 streams:
  RX P=0/Q=0 (socket 0) -> TX P=1/Q=0 (socket 0) peer=50:E3:44:47:9B:02
  RX P=1/Q=0 (socket 0) -> TX P=0/Q=0 (socket 0) peer=50:BE:76:14:22:01
Logical Core 2 (socket 0) forwards packets on 2 streams:
  RX P=0/Q=1 (socket 0) -> TX P=1/Q=1 (socket 0) peer=50:E3:44:47:9B:02
  RX P=1/Q=1 (socket 0) -> TX P=0/Q=1 (socket 0) peer=50:BE:76:14:22:01
Logical Core 3 (socket 0) forwards packets on 2 streams:
  RX P=0/Q=2 (socket 0) -> TX P=1/Q=2 (socket 0) peer=50:E3:44:47:9B:02
  RX P=1/Q=2 (socket 0) -> TX P=0/Q=2 (socket 0) peer=50:BE:76:14:22:01
Logical Core 4 (socket 0) forwards packets on 2 streams:
  RX P=0/Q=3 (socket 0) -> TX P=1/Q=3 (socket 0) peer=50:E3:44:47:9B:02
  RX P=1/Q=3 (socket 0) -> TX P=0/Q=3 (socket 0) peer=50:BE:76:14:22:01

  mac packet forwarding packets/burst=32
  nb forwarding cores=4 - nb forwarding ports=2
  port 0: RX queue number: 4 Tx queue number: 4
    Rx offloads=0x80000 Tx offloads=0x10000
    RX queue: 0
      RX desc=2048 - RX free threshold=32
      RX threshold registers: pthresh=0 hthresh=0  wthresh=0
      RX Offloads=0x0
    TX queue: 0
      TX desc=2048 - TX free threshold=32
      TX threshold registers: pthresh=0 hthresh=0  wthresh=0
      TX offloads=0x10000 - TX RS bit threshold=32
  port 1: RX queue number: 4 Tx queue number: 4
    Rx offloads=0x80000 Tx offloads=0x10000
    RX queue: 0
      RX desc=2048 - RX free threshold=32
      RX threshold registers: pthresh=0 hthresh=0  wthresh=0
      RX Offloads=0x0
    TX queue: 0
      TX desc=2048 - TX free threshold=32
      TX threshold registers: pthresh=0 hthresh=0  wthresh=0
      TX offloads=0x10000 - TX RS bit threshold=32
testpmd> 
2024/01/28 10:22:55 Clearing testpmd stats in VMI...
2024/01/28 10:22:55 Clearing Trex console stats before test...
2024/01/28 10:22:57 Running traffic for 1s...
2024/01/28 10:22:59 Monitoring traffic generator side drop rates every 10s during the test duration...
2024/01/28 10:23:01 GetGlobalStats JSON Response:
{
    "id": "vaxarlhy",
    "jsonrpc": "2.0",
    "result": {
        "m_active_flows": 0.0,
        "m_active_sockets": 0,
        "m_bw_per_core": 14.48548698425293,
        "m_cpu_util": 7.059671878814697,
        "m_cpu_util_raw": 15.0,
        "m_open_flows": 0.0,
        "m_platform_factor": 1.0,
        "m_rx_bps": 1932498944.0,
        "m_rx_core_pps": 0.0,
        "m_rx_cpu_util": 0.0,
        "m_rx_drop_bps": 0.0,
        "m_rx_pps": 3774411.75,
        "m_socket_util": 0.0,
        "m_total_alloc_error": 0,
        "m_total_clients": 0,
        "m_total_nat_active": 0,
        "m_total_nat_learn_error": 0,
        "m_total_nat_no_fid": 0,
        "m_total_nat_open": 0,
        "m_total_nat_syn_wait": 0,
        "m_total_nat_time_out": 0,
        "m_total_nat_time_out_wait_ack": 0,
        "m_total_queue_drop": 0,
        "m_total_queue_full": 0,
        "m_total_rx_bytes": 512000256,
        "m_total_rx_pkts": 8000004,
        "m_total_servers": 0,
        "m_total_tx_bytes": 544000272,
        "m_total_tx_pkts": 8000004,
        "m_tx_bps": 2045255680.0,
        "m_tx_cps": 0.0,
        "m_tx_expected_bps": 0.0,
        "m_tx_expected_cps": 0.0,
        "m_tx_expected_pps": 0.0,
        "m_tx_pps": 3759661.25
    }
}
2024/01/28 10:23:01 finished polling for drop rates
2024/01/28 10:23:01 traffic Generator Max Drop Rate: 0.000000Bps
2024/01/28 10:23:03 GetPortStats JSON Response:
{
    "id": "76og4wei",
    "jsonrpc": "2.0",
    "result": {
        "ibytes": 0,
        "ierrors": 0,
        "ipackets": 0,
        "m_cpu_util": 0.0,
        "m_total_rx_bps": 0.0,
        "m_total_rx_pps": 0.0,
        "m_total_tx_bps": 511313920.0,
        "m_total_tx_pps": 939915.3125,
        "obytes": 544000272,
        "oerrors": 0,
        "opackets": 8000004
    }
}
2024/01/28 10:23:05 GetPortStats JSON Response:
{
    "id": "7j4exe88",
    "jsonrpc": "2.0",
    "result": {
        "ibytes": 512000256,
        "ierrors": 0,
        "ipackets": 8000004,
        "m_cpu_util": 0.0,
        "m_total_rx_bps": 120781184.0,
        "m_total_rx_pps": 235900.734375,
        "m_total_tx_bps": 0.0,
        "m_total_tx_pps": 0.0,
        "obytes": 0,
        "oerrors": 0,
        "opackets": 0
    }
}
2024/01/28 10:23:05 traffic Generator port 0 Packet output errors: 0
2024/01/28 10:23:05 traffic Generator port 1 Packet output errors: 0
2024/01/28 10:23:05 traffic Generator packet sent via port 0: 8000004
2024/01/28 10:23:05 get testpmd stats in VM-Under-Test...
2024/01/28 10:23:07 testpmd stats:
show fwd stats all

  ------- Forward Stats for RX Port= 0/Queue= 2 -> TX Port= 1/Queue= 2 -------
  RX-packets: 4000002        TX-packets: 4000002        TX-dropped: 0             

  ------- Forward Stats for RX Port= 0/Queue= 3 -> TX Port= 1/Queue= 3 -------
  RX-packets: 4000002        TX-packets: 4000002        TX-dropped: 0             

  ---------------------- Forward statistics for port 0  ----------------------
  RX-packets: 8000004        RX-dropped: 0             RX-total: 8000004
  TX-packets: 0              TX-dropped: 0             TX-total: 0
  ----------------------------------------------------------------------------

  ---------------------- Forward statistics for port 1  ----------------------
  RX-packets: 0              RX-dropped: 0             RX-total: 0
  TX-packets: 8000004        TX-dropped: 0             TX-total: 8000004
  ----------------------------------------------------------------------------

  +++++++++++++++ Accumulated forward statistics for all ports+++++++++++++++
  RX-packets: 8000004        RX-dropped: 0             RX-total: 8000004
  TX-packets: 8000004        TX-dropped: 0             TX-total: 8000004
  ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
testpmd> 
2024/01/28 10:23:07 VMI-Under-Test's side packets Dropped: Rx: 0; TX: 0
2024/01/28 10:23:07 VMI-Under-Test's side test packets received (including dropped, excluding non-related packets): 8000004
2024/01/28 10:23:07 Trying to delete VMI: "dpdk-checkup-ns-1/vmi-under-test-pnjqb"
2024/01/28 10:23:07 Trying to delete VMI: "dpdk-checkup-ns-1/dpdk-traffic-gen-pnjqb"
2024/01/28 10:23:07 Deleting ConfigMap "dpdk-checkup-ns-1/dpdk-traffic-gen-config-pnjqb"...
2024/01/28 10:23:07 Deleting ConfigMap "dpdk-checkup-ns-1/vmi-under-test-config-pnjqb"...
2024/01/28 10:23:07 Waiting for VMI "dpdk-checkup-ns-1/vmi-under-test-pnjqb" to be deleted...
2024/01/28 10:23:22 VMI "dpdk-checkup-ns-1/vmi-under-test-pnjqb" was deleted successfully
2024/01/28 10:23:22 Waiting for VMI "dpdk-checkup-ns-1/dpdk-traffic-gen-pnjqb" to be deleted...
2024/01/28 10:23:22 VMI "dpdk-checkup-ns-1/dpdk-traffic-gen-pnjqb" was deleted successfully
```

orelmisan reviewed (approved):
Thank you

RamLavi commented:
Ran it a few more times to make sure the solution is stable. 
Merging.
